File: packages/core/src/models/task.ts
import { v4 as uuidv4 } from 'uuid';

export type TaskStatus = 'todo' | 'in_progress' | 'completed' | 'blocked';
export type EnergyLevel = 'low' | 'medium' | 'high';
export type PinType = 'soft' | 'hard';
export type RecurrenceFrequency = 'daily' | 'weekly' | 'monthly';

export interface Recurrence {
  frequency: RecurrenceFrequency | null;
  days_of_week: string[];  // ["Monday", "Wednesday"]
  time_of_day: string | null;  // "14:00"
  end_date: string | null;  // ISO date string
}

export interface Task {
  id: string;
  project_id: string;
  parent_task_id: string | null;
  title: string;
  description: string | null;
  priority: number;  // 1-5
  difficulty: number;  // 1-5
  energy_level: EnergyLevel;
  estimated_minutes: number;
  actual_minutes: number | null;
  status: TaskStatus;
  dependencies: string[];
  tags: string[];
  deadline: string | null;  // ISO datetime string
  scheduled_for: string | null;  // ISO date string
  created_at: string;  // ISO datetime string
  updated_at: string;  // ISO datetime string
  completed_at: string | null;  // ISO datetime string

  // Habit-specific fields
  is_habit: boolean;
  recurrence: Recurrence;
  is_paused: boolean;
  paused_until: string | null;  // ISO date string

  // Quick capture / inbox fields
  needs_review: boolean;

  // Pinning for drag & drop scheduling
  is_pinned: boolean;
  pin_type: PinType;
}

export interface TaskCreate {
  project_id: string;
  parent_task_id?: string | null;
  title: string;
  description?: string | null;
  priority?: number;
  difficulty?: number;
  energy_level?: EnergyLevel;
  estimated_minutes?: number;
  dependencies?: string[];
  tags?: string[];
  deadline?: string | null;
  scheduled_for?: string | null;
  is_habit?: boolean;
  recurrence?: Partial<Recurrence>;
  needs_review?: boolean;
  // Fixed time scheduling (meetings, appointments)
  is_pinned?: boolean;
  pin_type?: PinType;
}

export interface TaskUpdate {
  title?: string;
  description?: string | null;
  priority?: number;
  difficulty?: number;
  energy_level?: EnergyLevel;
  estimated_minutes?: number;
  actual_minutes?: number | null;
  status?: TaskStatus;
  dependencies?: string[];
  tags?: string[];
  deadline?: string | null;
  scheduled_for?: string | null;
  needs_review?: boolean;
  is_pinned?: boolean;
  pin_type?: PinType;
  is_paused?: boolean;
  paused_until?: string | null;
  recurrence?: Partial<Recurrence>;
}

export function createDefaultRecurrence(): Recurrence {
  return {
    frequency: null,
    days_of_week: [],
    time_of_day: null,
    end_date: null,
  };
}

export function createTask(data: TaskCreate): Task {
  const now = new Date().toISOString();
  return {
    id: uuidv4(),
    project_id: data.project_id,
    parent_task_id: data.parent_task_id ?? null,
    title: data.title,
    description: data.description ?? null,
    priority: data.priority ?? 3,
    difficulty: data.difficulty ?? 3,
    energy_level: data.energy_level ?? 'medium',
    estimated_minutes: data.estimated_minutes ?? 60,
    actual_minutes: null,
    status: 'todo',
    dependencies: data.dependencies ?? [],
    tags: data.tags ?? [],
    deadline: data.deadline ?? null,
    scheduled_for: data.scheduled_for ?? null,
    created_at: now,
    updated_at: now,
    completed_at: null,
    is_habit: data.is_habit ?? false,
    recurrence: {
      ...createDefaultRecurrence(),
      ...data.recurrence,
    },
    is_paused: false,
    paused_until: null,
    needs_review: data.needs_review ?? false,
    is_pinned: data.is_pinned ?? false,
    pin_type: data.pin_type ?? 'soft',
  };
}

File: taskman/server/models/task.py
from pydantic import BaseModel, Field
from datetime import datetime, date
from typing import Optional, List, Dict
from taskman.shared.constants import TaskStatus

class Recurrence(BaseModel):
    """Recurrence pattern for habits"""
    frequency: Optional[str] = None  # "daily", "weekly", "monthly"
    days_of_week: List[str] = Field(default_factory=list)  # ["Monday", "Wednesday"]
    time_of_day: Optional[str] = None  # "14:00"
    end_date: Optional[date] = None

class Task(BaseModel):
    id: str
    project_id: str
    parent_task_id: Optional[str] = None
    title: str
    description: Optional[str] = None
    priority: int = Field(ge=1, le=5, default=3)
    difficulty: int = Field(ge=1, le=5, default=3)
    energy_level: str = Field(default="medium")  # low, medium, high
    estimated_minutes: int = 60
    actual_minutes: Optional[int] = None
    status: TaskStatus = TaskStatus.TODO
    dependencies: List[str] = Field(default_factory=list)
    tags: List[str] = Field(default_factory=list)  # Hierarchical tags like ["Work/Meetings", "urgent"]
    deadline: Optional[datetime] = None
    scheduled_for: Optional[datetime] = None  # For hard-pinned tasks, includes date + time
    created_at: datetime = Field(default_factory=datetime.now)
    updated_at: datetime = Field(default_factory=datetime.now)
    completed_at: Optional[datetime] = None

    # Habit-specific fields
    is_habit: bool = False
    recurrence: Recurrence = Field(default_factory=Recurrence)
    is_paused: bool = False  # Whether the habit is paused (excluded from scheduling)
    paused_until: Optional[date] = None  # Auto-resume date (optional)

    # Quick capture / inbox fields
    needs_review: bool = False  # True if task needs user review (quick capture)

    # Pinning for drag & drop scheduling
    is_pinned: bool = False
    pin_type: str = Field(default="soft")  # "soft" or "hard"

    class Config:
        json_encoders = {
            datetime: lambda v: v.isoformat(),
            date: lambda v: v.isoformat()
        }

class TaskCreate(BaseModel):
    project_id: str
    parent_task_id: Optional[str] = None
    title: str
    description: Optional[str] = None
    priority: int = Field(ge=1, le=5, default=3)
    difficulty: int = Field(ge=1, le=5, default=3)
    energy_level: str = "medium"
    estimated_minutes: int = 60
    dependencies: List[str] = Field(default_factory=list)
    tags: List[str] = Field(default_factory=list)
    deadline: Optional[datetime] = None
    scheduled_for: Optional[datetime] = None  # For hard-pinned tasks, includes date + time

    # Habit fields
    is_habit: bool = False
    recurrence: Optional[Recurrence] = None

    # Quick capture
    needs_review: bool = False

class TaskUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    priority: Optional[int] = Field(None, ge=1, le=5)
    difficulty: Optional[int] = Field(None, ge=1, le=5)
    energy_level: Optional[str] = None
    estimated_minutes: Optional[int] = None
    actual_minutes: Optional[int] = None
    status: Optional[TaskStatus] = None
    dependencies: Optional[List[str]] = None
    tags: Optional[List[str]] = None
    deadline: Optional[datetime] = None
    scheduled_for: Optional[datetime] = None  # For hard-pinned tasks, includes date + time
    needs_review: Optional[bool] = None
    is_pinned: Optional[bool] = None
    pin_type: Optional[str] = None
    is_paused: Optional[bool] = None
    paused_until: Optional[date] = None
    recurrence: Optional[Recurrence] = None  # For updating habit schedule